### PR TITLE
EE-7726: Profiling events to correlate flow component executions with thread and thread jumps.

### PR DESCRIPTION
**Pull request by fsgonz**

### CODE REVIEW

**Review by IvanAndresFritzler — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/mule/runtime/core/api/profiling/notification/ProcessingStrategyProfilingEventType.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ * @since 4.4
+ */
+@Experimental
+public class ProcessingStrategyProfilingEventType implements ProfilingEventType<ProcessingStrategyProfilingEventContext> {
+
+  private final String profilingEventTypeIdentifier;
+  private final String profilingEventTypeNamespace;
+
+  // TODO Can this be refactored and become a generic ProfilingEventType generator?
```
[the remark this answers is not in the document]
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/api/profiling/notification/ProcessingStrategyProfilingEventType.java:29]: I've tried to create a ProfilingEventType generator but this will result in returning a generic profiling event type. So in that case for compile-time type verification a cast will be needed or no compile time type verification for the profiling event context will be done. 
Not sure if this is appropriate. Create a ticket for technical debt for review.

**Comment by fsgonz** [comment]
--validate

**Comment by fsgonz** [comment]
We'll add more tests.

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/PipelineProcessingStrategyReactiveProcessorBuilder.java [hunk 2]
```diff
@@ -31,56 +38,80 @@
...
+        .profileEvent(location, startingFlowExecutionDataproducer, artifactId, artifactType)
         .doOnSubscribe(subscription -> currentThread().setContextClassLoader(executionClassloader))
-        .transform(pipeline);
+        .transform(pipeline)
+        .profileEvent(location, flowExecutedDataProducer, artifactId, artifactType);
+  }
+
+  private Optional<ProfilingDataProducer<ProcessingStrategyProfilingEventContext>> dataProducerFromProfilingService(
+                                                                                                                    ProfilingEventType<ProcessingStrategyProfilingEventContext> profilingEventType) {
+    return ofNullable(profilingService).map(profilingService -> profilingService.getProfilingDataProducer(profilingEventType));
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/PipelineProcessingStrategyReactiveProcessorBuilder.java:114]: Using the optinal is overkill here for just a null check

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java [hunk 3]
```diff
@@ -132,6 +139,24 @@ private MonoBuilder(Mono<CoreEvent> mono) {
...
+    @Override
+    public ReactorPublisherBuilder<Mono<CoreEvent>> profileEvent(ComponentLocation location,
+                                                                 Optional<? extends ProfilingDataProducer> dataProducer,
+                                                                 String artifactId, String artifactType) {
+      mono =
+          dataProducer
+              .map(dp -> mono.doOnNext(e -> dp
+                  .triggerProfilingEvent(new ComponentProcessingStrategyProfilingEventContext(e, location,
+                                                                                              Thread
+                                                                                                  .currentThread()
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java:151]: static import

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java [hunk 4]
```diff
@@ -180,5 +205,21 @@ private FluxBuilder(Flux<CoreEvent> flux) {
...
+                                                                                new ComponentProcessingStrategyProfilingEventContext(e,
+                                                                                                                                     location,
+                                                                                                                                     Thread
+                                                                                                                                         .currentThread()
+                                                                                                                                         .getName(),
+                                                                                                                                     artifactId,
+                                                                                                                                     artifactType,
+                                                                                                                                     currentTimeMillis()))))
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/reactor/builder/ReactorPublisherBuilder.java:220]: extract the inside of doOnNext that is the same as for the mono

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+import org.mule.runtime.api.component.location.ComponentLocation;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.core.api.processor.ReactiveProcessor;
+import org.mule.runtime.core.internal.processor.chain.InterceptedReactiveProcessor;
+import org.mule.runtime.core.privileged.processor.chain.HasLocation;
+
+/**
+ * Utility methods for handling profiling of processing strategy.
+ * <p>
+ * MULE-19594: Handle expression problem for Reactive Processor (ComponentLocation)
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java:20]: what does this mean?

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  /**
+   * @param muleContext the {@link MuleContext} from which to return the artifactId
+   * @return the artifact id if possible.
+   */
+  public static String getArtifactId(MuleContext muleContext) {
+    if (muleContext.getConfiguration() == null) {
+      return "UNKNOWN_ARTIFACT_ID";
+    }
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java:53]: Can rthis happen at all?

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  /**
+   * @param muleContext the {@link MuleContext} from which to return the artifactId
+   * @return the artifact id if possible.
+   */
+  public static String getArtifactType(MuleContext muleContext) {
+    if (muleContext.getArtifactType() == null) {
+      return "UNKNOWN_ARTIFACT_TYPE";
+    }
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java:65]: Can rthis happen at all?

File: core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingDataConsumerDiscoveryStrategy.java [hunk 8]
```diff
@@ -7,25 +7,26 @@
...
  * @since 4.4
  */
-public class DefaultProfilerDataConsumerDiscoveryStrategy implements ProfilerDataConsumerDiscoveryStrategy {
+public class DefaultProfilingDataConsumerDiscoveryStrategy implements ProfilingDataConsumerDiscoveryStrategy {
 
   @Override
   public <S extends ProfilingDataConsumer<T>, T extends ProfilingEventContext> Set<S> discover() {
-    // No data consumers.
-    return emptySet();
+    return (Set<S>) ImmutableSet.of(new LoggerComponentProcessingStrategyDataConsumer());
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/DefaultProfilingDataConsumerDiscoveryStrategy.java:29]: USe. `Collections.singleton` instead

File: core/src/main/java/org/mule/runtime/core/internal/profiling/consumer/LoggerComponentProcessingStrategyDataConsumer.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+
+  public static final String PROFILING_EVENT_TIMESTAMP_KEY = "profilingEventTimestamp";
+  public static final String PROCESSING_THREAD_KEY = "processingThread";
+  public static final String ARTIFACT_ID_KEY = "artifactId";
+  public static final String ARTIFACT_TYPE_KEY = "artifactType";
+  public static final String RUNTIME_CORE_EVENT_CORRELATION_ID = "runtimeCoreEventCorrelationId";
+  public static final String PROFILING_EVENT_TYPE = "profilingEventType";
+  public static final String LOCATION = "location";
+
+  private final Gson gson = new Gson();
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/profiling/consumer/LoggerComponentProcessingStrategyDataConsumer.java:49]: is this Gson threwad safe?

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/HasLocation.java [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+package org.mule.runtime.core.privileged.processor.chain;
+
+import org.mule.runtime.api.component.location.ComponentLocation;
+import org.mule.runtime.core.api.processor.ReactiveProcessor;
+
+/**
+ * An interface that indicates that the {@link ComponentLocation} can be resolved. It is used as a WA so that the subtypes of
+ * {@link ReactiveProcessor} that are associated to a location can be identified and a location can be obtained for those types.
+ *
+ * // TODO MULE-19594: Handle expression problem for Reactive Processor (ComponentLocation)
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/chain/HasLocation.java:17]: what does this mean?

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/HasLocation.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+ */
+public interface HasLocation {
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/chain/HasLocation.java:19]: `@NoImplement`? why is this needed as privleged?

File: tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java [hunk 12]
```diff
@@ -22,6 +22,7 @@
 import static org.mule.runtime.config.internal.error.MuleCoreErrorTypeRepository.MULE_CORE_ERROR_TYPE_REPOSITORY;
 import static org.mule.runtime.core.api.config.MuleManifest.getProductVersion;
 import static org.mule.runtime.core.api.config.MuleProperties.OBJECT_STORE_MANAGER;
+import static org.mule.runtime.core.api.config.bootstrap.ArtifactType.APP;
```
Comment (elrodro83) [inline tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java:25]: ?

File: core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamEmitterProcessingStrategyTestCase.java [hunk 13]
```diff
@@ -100,11 +111,32 @@
...
-  public ProactorStreamEmitterProcessingStrategyTestCase(Mode mode) {
-    super(mode);
+  public ProactorStreamEmitterProcessingStrategyTestCase(Mode mode, boolean profiling) {
+    super(mode, profiling);
+  }
+
+  @Override
+  protected Map<String, Object> getStartUpRegistryObjects() {
+    Map<String, Object> startupRegistryObjects = new HashMap<>();
+    startupRegistryObjects.put(MuleProperties.MULE_PROFILING_SERVICE_KEY, profilingService);
```
Comment (elrodro83) [inline core-tests/src/test/java/org/mule/runtime/core/internal/processor/strategy/ProactorStreamEmitterProcessingStrategyTestCase.java:138]: static import

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java (hunk 12 above)
Comment (fsgonz) [reply tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java:25]: Needed so that the muleContext is mocked with an ArtifactType,

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java (hunk 5 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java:20]: Changed to "Problem for retrieving component location from generic reactive processor." 
https://en.wikipedia.org/wiki/Expression_problem

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java (hunk 6 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/internal/processor/strategy/util/ProfilingUtils.java:53]: Meant to be defensive. Removed.

File: core/src/main/java/org/mule/runtime/core/internal/profiling/consumer/LoggerComponentProcessingStrategyDataConsumer.java (hunk 9 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/internal/profiling/consumer/LoggerComponentProcessingStrategyDataConsumer.java:49]: Yes. https://www.javadoc.io/doc/com.google.code.gson/gson/2.8.0/com/google/gson/Gson.html

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/HasLocation.java (hunk 10 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/privileged/processor/chain/HasLocation.java:17]: Changed to refactor the way of retrieving the component location from generic reactive processor.

File: tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java (hunk 12 above)
Comment (fsgonz) [reply tests/unit/src/main/java/org/mule/tck/util/MuleContextUtils.java:25]: Added a missing line of code. :(

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/HasLocation.java (hunk 11 above)
Comment (fsgonz) [reply core/src/main/java/org/mule/runtime/core/privileged/processor/chain/HasLocation.java:19]: No. Moved.

File: core/src/main/java/org/mule/runtime/core/internal/processor/strategy/TransactionAwareStreamEmitterProcessingStrategyFactory.java [hunk 14]
```diff
@@ -6,6 +6,10 @@
  */
 package org.mule.runtime.core.internal.processor.strategy;
 
+import static org.mule.runtime.core.api.lifecycle.LifecycleUtils.*;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/internal/processor/strategy/TransactionAwareStreamEmitterProcessingStrategyFactory.java:9]: no & imports

File: core/src/main/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainBuilder.java [hunk 15]
```diff
@@ -25,8 +25,7 @@
 import org.mule.runtime.core.api.event.CoreEvent;
 import org.mule.runtime.core.api.exception.FlowExceptionHandler;
 import org.mule.runtime.core.api.exception.NullExceptionHandler;
-import org.mule.runtime.core.api.processor.InterceptingMessageProcessor;
-import org.mule.runtime.core.api.processor.Processor;
+import org.mule.runtime.core.api.processor.*;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/privileged/processor/chain/DefaultMessageProcessorChainBuilder.java:28]: *

File: core/src/main/java/org/mule/runtime/core/api/processor/HasLocation.java [hunk 16]
```diff
@@ -5,17 +5,18 @@
  * LICENSE.txt file.
  */
 
-package org.mule.runtime.core.privileged.processor.chain;
+package org.mule.runtime.core.api.processor;
```
Comment (elrodro83) [inline core/src/main/java/org/mule/runtime/core/api/processor/HasLocation.java:8]: why do you need this in api? the uses in this pr are internal

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Review by elrodro83 — APPROVED** [review]

**Review by IvanAndresFritzler — APPROVED** [review]
